Define unspecified behavior (#214)
rust-lang/unsafe-code-guidelines: reference/src/glossary.md

Also see: [Soundness][soundness].

#### Unspecified behavior
[unspecified]: #unspecified

> *Unspecified behavior* is not an error condition in the abstract machine, but beyond that, the Rust language provides no other guarantees about what behavior these programs have.

Reviewer: This is an odd introduction... the first thing you say is what this *not* is.

Also, we are moving towards a more general "assumptions made by compiler" def.n for UB.

Proposal:

> *Unspecified behavior* is behavior of the Rust Abstract Machine that the Rust language provides no guarantee for. Unspecified behavior always comes with a set of behaviors that the implementation can pick from.

The latter part is important. I don't think "anything but the error state" is a useful spec. And for your example of field offsets, there is such a set: In https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/layout/structs-and-tuples.md we define what the dregrees of freedom are here for the compiler.

Reviewer: > "the behavior for which the Rust Abstract Machine provides no guarantees for" is undefined behavior

Not really -- there's no behavior in UB for the R-AM, it is an error state.

> The only thing we guarantee about unspecified behavior is that it is not an error in the abstract machine.

That's useless. Then the behavior could still be "replace all memory contents by 0x00", making it impossible to program.

We *always* need to give a bound on what "unspecified behavior" can do, or we might as well declare it UB.

Reviewer: > *Unspecified behavior* occurs in the Rust language when the implementation is free to pick any one of a given set of possible behaviors of the R-AM. The implementation does not have to document that choice nor commit to it, and the choice it makes can vary even within the execution of a single program.

Author: > That's useless. Then the behavior could still be "replace all memory contents by 0x00", making it impossible to program.
>
> We always need to give a bound on what "unspecified behavior" can do, or we might as well declare it UB.

This definition isn't useless since it provides a guarantee over undefined behavior. Text that uses it might provide extra bounds, but I don't think this definition needs to try to provide such bounds nor require them to exist.

Reviewer: It doesn't provide any useful guarantee over UB. It is literally impossible to write a guaranteed UB-free program with your proposed definition. The moment an unspecified operation occurs, *you have no idea what the state of the R-AM is*. So the next operation is almost certainly UB under some possible choices for what the unspecified operation does.

So, I feel very strongly that "unspecified" only makes any sense when paired with a set of options to choose from, and that we should define it in a way that mandates this.

Author: > The moment an unspecified operation occurs, you have no idea what the state of the R-AM is. 

That's sometimes the case, which is why miri cannot model everything.

Author: I mean, FFI is a prime example, what does calling an `extern "C"` function do ? It calls it according to some ABI, right? What if there is no spec for such an ABI, or the platform doesn't have a C ABI at all? The compiler and implementation can still make things work, e.g., clang and rustc can agree on an ABI, and such agreement can change with time. 

But how would you "bound" that behavior in wording ?

Reviewer: To elaborate on "useless", imagine we had a function `fn unspec();` and said calling it is unspecified behavior. "Anything except for UB can happen". Well, one possible choice for "anything" is "oops your memory is empty now, we deallocated all of it", so the following program could have UB:
```rust
let x = 4;
unspec();
assert!(x == 4); // UB! x might not be allocated any more
```
We could carefully try to restrict what unspecified behavior can do *in general*, but that's going to be super painful. So unless there is a strong motivation for having "(almost) unbounded unspecified behavior", we better avoid it.

Author: Unless you allow "bounded by a potentially unspecified bound" I'm not sure how you can describe FFI.

Author: But at that point, I'd rather leave the bound to whatever feature decides to be "unspecified behavior", and if that feature decides to provide no bounds, and that in your opinion makes the feature useless, then just make the case against adding such a feature to the language?

Reviewer: FFI isn't "unspecified behavior"... that would be rather catastrophic as one couldn't program against it.^^ Specifying FFI is hard, but we shouldn't pretend that we can properly handle it by saying "unspecified". We need to define cross-language linking to specify FFI. Without xLTO we could do it on the target/assembly level; with xLTO... TBH at that point we probably have to work on the LLVM level as I doubt we can make C programs run on the R-AM.^^

OTOH, struct field offsets are a good example for unspecified behavior precisely because we can bound the choices but do not want to commit. *That's* what we should use it for; not as an excuse for "sorry it's hard we don't know what to say".

Author: > FFI isn't "unspecified behavior"... that would be rather catastrophic as one couldn't program against it.^^

Really? What is it then ?

> For example, the field offsets of `repr(Rust)` types is *unspecified*, and a Rust program that prints the offset of a field exhibits *unspecified behavior*: it prints somethingbut we do not make any guarantees about what it prints. This might change across compiler versions, depending on compiler flags, or even across compiler invocations.

Reviewer 2: *something but

Author: ```suggestion
For example, the field offsets of `repr(Rust)` types is *unspecified*, so a Rust program that prints the offset of a field exhibits *unspecified behavior*: it prints something, but we do not make any guarantees about what it prints. What it prints can therefore change across compiler versions, depending on compiler flags, or even across compiler invocations.
```

> Programs that make assumptions about what a particular source of *unspecified behavior* does often end up exhibiting *undefined behavior* when those assumptions are incorrect. For example, making an assumption that a field of a `repr(Rust)` struct is at a particular offset that is invalidated when that offset changes.

Author: ```suggestion
Programs that make assumptions about what a particular source of *unspecified behavior* does often end up exhibiting *undefined behavior* when those assumptions are incorrect. For example, making the assumption that a field of a `repr(Rust)` struct is at a particular offset might lead a program to exhibit *undefined behavior* if that assumption is incorrect.
```

Author: Reword

> Programs that make assumptions about what a particular source of *unspecified behavior* does often end up exhibiting *undefined behavior* when those assumptions are silently invalidated. For example, making an assumption that a field of a `repr(Rust)` struct is at a particular offse that is invalidated when that offset changes.

Author: ```suggestion
Programs that make assumptions about what a particular source of *unspecified behavior* does often end up exhibiting *undefined behavior* when those assumptions are silently invalidated. For example, making an assumption that a field of a `repr(Rust)` struct is at a particular offset that is invalidated when that offset changes.
```

> Programs that make assumptions about what a particular source of *unspecified behavior* does often end up exhibiting *undefined behavior* when those assumptions are silently invalidated. For example, making an assumption that a field of a `repr(Rust)` struct is at a particular offset that is invalidated when that offset changes.

Author: ```suggestion
Programs that make assumptions about what a particular source of *unspecified behavior* does often end up exhibiting *undefined behavior* when those assumptions are incorrect. For example, making an assumption that a field of a `repr(Rust)` struct is at a particular offset that is invalidated when that offset changes.
```
